cmd/otel-allocator/main.go

```diff
@@ -0,0 +1,156 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/fsnotify/fsnotify"
+	gokitlog "github.com/go-kit/log"
+	"github.com/gorilla/mux"
+
+	"github.com/otel-allocator/allocation"
+
+	"github.com/otel-allocator/collector"
+	"github.com/otel-allocator/config"
+	lbdiscovery "github.com/otel-allocator/discovery"
+)
+
+const (
+	configDir  = "/conf/"
+	listenAddr = ":8080"
+)
+
+func main() {
+	ctx := context.Background()
+
+	// watcher to monitor file changes in ConfigMap
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatalf("Can't start the watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	if err := watcher.Add(configDir); err != nil {
+		log.Fatalf("Can't add directory to watcher: %v", err)
+	}
+
+	srv, err := newServer(listenAddr)
+	if err != nil {
+		log.Fatalf("Can't start the server: %v", err)
+	}
+
+	interrupts := make(chan os.Signal, 1)
+	signal.Notify(interrupts, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		if err := srv.Start(); err != http.ErrServerClosed {
+			log.Fatalf("Can't start the server: %v", err)
+		}
+	}()
+
+	for {
+		select {
+		case <-interrupts:
+			if err := srv.Shutdown(ctx); err != nil {
+				log.Println(err)
+				os.Exit(1)
+			}
+			os.Exit(0)
+		case event := <-watcher.Events:
+			switch event.Op {
+			case fsnotify.Create:
+				log.Println("ConfigMap updated!")
+				// Restart the server to pickup the new config.
+				if err := srv.Shutdown(ctx); err != nil {
+					log.Fatalf("Cannot shutdown the server: %v", err)
+				}
+				srv, err = newServer(listenAddr)
+				if err != nil {
+					log.Fatalf("Error restarting the server with new config: %v", err)
+				}
+				go func() {
+					if err := srv.Start(); err != http.ErrServerClosed {
+						log.Fatalf("Can't restart the server: %v", err)
+					}
+				}()
+			}
+		case err := <-watcher.Errors:
+			log.Printf("Watcher error: %v", err)
+		}
+	}
+}
+
+type server struct {
+	allocator        *allocation.Allocator
+	discoveryManager *lbdiscovery.Manager
+	k8sClient        *collector.Client
+	server           *http.Server
+}
+
+func newServer(addr string) (*server, error) {
+	allocator, discoveryManager, k8sclient, err := newAllocator(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	s := &server{
+		allocator:        allocator,
+		discoveryManager: discoveryManager,
+		k8sClient:        k8sclient,
+	}
+	router := mux.NewRouter()
+	router.HandleFunc("/jobs", func(w http.ResponseWriter, r *http.Request) {
+		allocation.JobHandler(w, r, allocator)
```

Comment:
Your `server` struct could actually provide two HTTP handler functions, like:

```go
func (*server) JobHandler(w http.ResponseWriter, r *http.Request)
```

This way, you can just pass `s.JobHandler` as a parameter to the router.